File: src/backup.rs
//! PATH backup management functionality.
//!
//! This module provides functionality for:
//! - Creating timestamped backups of PATH
//! - Listing backup history
//! - Managing backup storage location

use chrono::Local;
use serde::{Deserialize, Serialize};
use std::env;
use std::fs::{self, File};
use std::path::PathBuf;

/// Represents a PATH backup with timestamp
#[derive(Serialize, Deserialize)]
struct Backup {
    /// Timestamp when backup was created
    timestamp: String,
    /// Complete PATH string at backup time
    path: String,
}

/// Creates a new backup of the current PATH
///
/// # Returns
///
/// Returns std::io::Result indicating success or failure
///
/// # Example
///
/// ```
/// match backup::create_backup() {
///     Ok(_) => println!("Backup created successfully"),
///     Err(e) => eprintln!("Failed to create backup: {}", e),
/// }
/// ```
pub fn create_backup() -> std::io::Result<()> {
    let timestamp = Local::now().format("%Y%m%d%H%M%S").to_string();
    let path = env::var("PATH").unwrap_or_default();

    let backup = Backup {
        timestamp: timestamp.clone(),
        path,
    };

    let backup_dir = get_backup_dir();
    fs::create_dir_all(&backup_dir)?;

    let backup_file = backup_dir.join(format!("backup_{}.json", timestamp));
    let file = File::create(backup_file)?;
    serde_json::to_writer_pretty(file, &backup)?;

    Ok(())
}

/// Displays the history of PATH backups
///
/// Lists all available backups in chronological order
pub fn show_history() {
    let backup_dir = get_backup_dir();

    match fs::read_dir(&backup_dir) {
        Ok(entries) => {
            println!("Available backups:");
            for entry in entries.flatten() {
                println!("- {}", entry.file_name().to_string_lossy());
            }
        }
        Err(_) => {
            println!("No backups found.");
        }
    }
}

/// Gets the directory where backups are stored
///
/// # Returns
///
/// Returns a PathBuf pointing to the backup directory
pub fn get_backup_dir() -> PathBuf {
    let home_dir = dirs_next::home_dir().unwrap_or_else(|| PathBuf::from("/"));
    home_dir.join(".pathfinder_backups")
}

File: internal/backup/backup.go
package backup

import (
	"fmt"
	"os"
	"path/filepath"
	"time"
)

// BackupFormat represents the format for the backup
type BackupFormat string

const (
	FormatJSON BackupFormat = "json"
	FormatText BackupFormat = "text"
	FormatTOML BackupFormat = "toml"
)

// PathBackup represents a backup of the PATH environment variable
type PathBackup struct {
	Timestamp time.Time
	Entries   []string
	Format    BackupFormat
}

// GetBackupDir returns the directory where backups are stored
func GetBackupDir() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", fmt.Errorf("failed to get home directory: %w", err)
	}
	
	backupDir := filepath.Join(home, ".pathmaster", "backups")
	
	// Create the directory if it doesn't exist
	if err := os.MkdirAll(backupDir, 0755); err != nil {
		return "", fmt.Errorf("failed to create backup directory: %w", err)
	}
	
	return backupDir, nil
}

// CreateBackup creates a new backup of the current PATH
func CreateBackup(entries []string, format BackupFormat) (*PathBackup, error) {
	backup := &PathBackup{
		Timestamp: time.Now(),
		Entries:   entries,
		Format:    format,
	}
	
	// This is a placeholder. In the full implementation, we would:
	// 1. Serialize the backup based on the format
	// 2. Write it to the backup directory with a timestamp-based filename
	
	return backup, nil
}

// ListBackups returns a list of available backups
func ListBackups() ([]*PathBackup, error) {
	// This is a placeholder. In the full implementation, we would:
	// 1. Scan the backup directory
	// 2. Parse each backup file
	// 3. Return a list of PathBackup objects
	
	return []*PathBackup{}, nil
}